hypothesis-python/src/hypothesis/internal/conjecture/pareto.py

```diff
@@ -0,0 +1,269 @@
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+from __future__ import absolute_import, division, print_function
+
+from enum import Enum
+
+from sortedcontainers import SortedList
+
+from hypothesis.internal.conjecture.data import ConjectureData, ConjectureResult, Status
+from hypothesis.internal.conjecture.junkdrawer import LazySequenceCopy, swap
+from hypothesis.internal.conjecture.shrinker import sort_key
+
+
+class DominanceRelation(Enum):
+    NO_DOMINANCE = 0
+    EQUAL = 1
+    LEFT_DOMINATES = 2
+    RIGHT_DOMINATES = 3
+
+
+def dominance(left, right):
+    """Returns the dominance relation between ``left`` and ``right``, according
+    to the rules that something dominates if and only if it is better in every
+    way.
+
+    The things we currently consider to be "better" are:
+
+        * Something that is smaller in shrinking order is better.
+        * Something that has higher status is better.
+        * Each ``interesting_origin`` is treated as its own score, so if two
+          interesting examples have different origins then neither dominates
+          the other.
+        * For each target observation, a higher score is better.
+
+    In "normal" operation where there are no bugs or target observations, the
+    pareto front only has one element (the smallest valid test case), but for
+    more structured or failing tests it can be useful to track, and future work
+    will depend on it more."""
+
+    if left.buffer == right.buffer:
+        return DominanceRelation.EQUAL
+
+    if sort_key(right.buffer) < sort_key(left.buffer):
+        result = dominance(right, left)
+        if result == DominanceRelation.LEFT_DOMINATES:
+            return DominanceRelation.RIGHT_DOMINATES
+        else:
+            # Because we have sort_key(left) < sort_key(right) the only options
+            # are that right is better than left or that the two are
+            # incomparable.
+            assert result == DominanceRelation.NO_DOMINANCE
+            return result
+
+    # Either left is better or there is no dominance relationship.
+    assert sort_key(left.buffer) < sort_key(right.buffer)
+
+    # The right is more interesting
+    if left.status < right.status:
+        return DominanceRelation.NO_DOMINANCE
+
+    # Things that are interesting for different reasons are incomparable in
+    # the dominance relationship.
+    if (
+        left.status == Status.INTERESTING
+        and left.interesting_origin != right.interesting_origin
+    ):
+        return DominanceRelation.NO_DOMINANCE
+
+    for target, score in left.target_observations.items():
+        if (
+            target in right.target_observations
+            and right.target_observations[target] > score
+        ):
+            return DominanceRelation.NO_DOMINANCE
```

Comment:
I'd find this more obvious with an explicit intersection:
```python
for target in set(left.target_observations).intersection(right.target_observations):
    if left.target_observations[target] < right.target_observations[target]:
        return DominanceRelation.NO_DOMINANCE
```
but a comment here (and ideally in the docstring too) would be sufficient - it just took me a bit of staring to confirm that targets are only relevant if they were observed on both sides.

Reply:
Hmm. Actually I think this code is wrong. I'll write some careful tests for it.

The intended logic is that target observations on either side are relevant, but that missing scores are treated as `-infinity`, but this misses the possibility of scores that are present on the RHS but not the left hand side.